### ISSUE TITLE
Append data to exsiting sas7bdat file

### ISSUE THREAD

**Issue by jinwzhang** [sassoftware/saspy#293]
Hi,
The issue is, I want to cowork with other collegues who are using sas. My team had a large log table saving daily log data (let's name it TableA).
Now I am calculating the same data using pandas and want to append my result into TableA. Is there any method I could do so with SASPy without truncating the TableA?
Thanks!

**Comment by tomweber-sas** [comment]
Hey, I don't have an append method in saspy, at the moment, for doing this. However, I bet I can before the end of the day :)

This seems like a very useful method. There are a coupe ways to do this in SAS, one eas, way (but 2 steps), and one error prone way (one step). The straight forward, safe way would be to create a data set from the new dataframe and proc append it to the existing SAS data set. The other would be to use a data step modify for the existing data set, writing out the rows from the dataframe directly. 

Considering all of the ways something could go wrong trying to directly modify the original data set, with data from the dataframe, my inclination is to create a temp dataset from the dataframe and then proc append it to the original data set, cleaning up the temp data after.

You can do this right now, yourself with saspy, but it's 2 steps you do manually. I can add an append method to the SASdata object, or an append_to= option on df2sd() since it's a dataframe you are starting with. I think the append method on SASdata would be cleaner, rather than muddying up df2sd, actually. Let me prototype something and we'll see what you think.

Tom

**Comment by tomweber-sas** [comment]
I've added an append (and delete) method to the SASdata object. I've pushed this to master so you can pull it and try it out, see what you think. It takes either a dataframe or a SASdata object and appends it to the data object. If a dataframe, I just call df2sd() on it first. For instance,
```
TableA = sas.sasdata('logtable')

df = pandas.dataframe() # whatever df you created containing compatible columns to TableA so it can append

TableA.append(df)

# or if you need to load a df with extra options on df2sd, or just have an other sas data  set
TableB = sas.df2sd(df, ...)
or
TableB = sas.sasdata('tableb')

TableA.append(TableB)

```

Proc Append will validate the data sets are compatible, so that's taken care of (they need to be).
If there are things like column length differences, the Proc has a FORCE option to force the append. I've added that as an optional parameter (False by default, of course).  

Give this a try and see what you think. I've still got more to do to validate it with other cases and such, but it works enough for you to see what you think or if there's anything else you need.

My little test case is as follows; first things I've tried. So, you can run that too or use it as a template for your test cases.  Let me know what you think!

Thanks,
Tom
```

import saspy
sas = saspy.SASsession(cfgname='iomj')
sas

cars = sas.sasdata('cars','sashelp')

df = cars.head()
df2 = cars.tail()

wcars  = sas.df2sd(df,  'wcars')
wcars2 = sas.df2sd(df2, 'wcars')

log = wcars.append(df2)
wcars.head(200)

log = wcars2.append(df)
wcars2.head(200)

log = wcars.append(df2, force=True)
wcars.head(200)

log = wcars2.append(df2, force=True)
wcars2.head(200)

log = wcars.append(cars)
wcars.head(2000)

log = wcars.append(cars, force=True)
wcars.head(2000)
```

**Comment by jinwzhang** [comment]
Hi Tom, 
Thanks for you kind update.
When I am testing the new package. (Download from git master as zip today,  and installed in offline server)
When I tried the following:
>>> type(car1)
<class 'saspy.sasdata.SASdata'>
>>> type(cars)
<class 'pandas.core.frame.DataFrame'>
>>> car1.append(cars)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'SASdata' object has no attribute 'append'

Excuse me for this silly question, maybe just let me know the correct way to update the pakage.

The additional question I had mentioned in a closed issue is:
The df2sd method could't really save the sas file physically. Even with the matching libref. Any suggestion?  see <Write out sas7bdat file #280>

**Comment by tomweber-sas** [comment]
yes, that doesn't appear that you installed the version from master. How are you installing it using pip?
Offline server? Meaning it can't access the internet? So you downloaded the zip to that machine. If so,
then when you unzip you should have the full directory structure of the repo. you should change directory to the top dir; it will have the setup.py file in it. Then the best way is to uninstall the currect version and then install this; I suspect if you look at the log from the install you tries, that it will show it didn't do anything because the existing version satisfied the request, or something like that.
The following should work:
```
cd to top dir you unzipped
pip uninstall -y saspy
pip install .
```
Give that a try and see if you then have the append method.

**Comment by jinwzhang** [comment]
Hey Tom, yes you are right, I think it was due to old version issue caused the problem. The Append method worked after uninstall and reinstall.

Somehow the sd2df( )  seems only worked for English version, after I swiched to Chinese GBK version. The method ends up with showing encoding error.
In the config file, I used:
ssh48      = {'saspath' : '/opt/sas94/SASHome/SASFoundation/9.4/bin/sas_zh',
            'ssh'     : '/usr/bin/ssh',
            'host'    : 'odsuser@XX.XXX.XXX.48',
            'encoding': 'gb2312',
            'options' : ["-fullstimer"]
            }
I also tried encoding as 'GBK', 'euc-cn' but sill report error. (just checked the encoding in SAS proc which shows 'euc-cn')

Error message (after google translate):
ERROR 24-2: Invalid value for "encoding" option.

ERROR: Error in FILENAME statement.
77 data _null_; set sashelp. \ 'Cars \' n;
78 format \ 'MSRP \' n best32 .; format \ 'Invoice \' n best32 .; format \ 'EngineSize \' n best32 .; format \ 'Cylinders \' n best32 .; format
78! \ 'Horsepower \' n best32 .; format \ 'MPG_City \' n best32 .;
79 format \ 'MPG_Highway \' n best32 .; format \ 'Weight \' n best32 .; format \ 'Wheelbase \' n best32 .; format \ 'Length \' n best32 .; file sock; put
79! \ 'Make \' n \ '02 \ 'x;
80 put \ 'Model \' n \ '02 \ 'x; put \' Type \ 'n \ '02 \' x; put \ 'Origin \' n \ '02 \ 'x; put \' DriveTrain \ 'n \ '02 \ 'x; put \' MSRP \ 'n \ '02 \' x; put \ 'Invoice \' n \ '02 \ 'x;
80! Put \ 'EngineSize \' n \ '02 \ 'x; put \' Cylinders \ 'n \ '02 \' x; put \ 'Horsepower \' n \ '02 \ 'x; put \' MPG_City \ ' n \ '02 \ 'x;
81 put \ 'MPG_Highway \' n \ '02 \ 'x; put \' Weight \ 'n \ '02 \' x; put \ 'Wheelbase \' n \ '02 \ 'x; put \' Length \ 'n \ '01 \ 'x; run;
NOTE: FILE SOCK is:
      File name = / home / odsuser / sock.dat,
      Owner name = odsuser, group name = odsuser,
      Access permissions = -rw-rw-r--,
      Last modified time = April 22, 2020 14:34:24

NOTE: There are 6420 records written to FILE SOCK.
      The minimum record length is 3.
      The maximum record length is 41.
NOTE: 428 observations were read from the data set SASHELP.CARS.
NOTE: Time used by "DATA Statement" (total processing time):
      Actual time 0.00 seconds
      User CPU time 0.00 seconds
      System CPU time 0.00 seconds
      Memory 558.53k
      OS memory 9124.00k
      Timestamp 2020-04-22 02:34:24 PM
      Step Count 11 Switch Count 0
      Number of page faults 0
      Page Recycling 45
      Number of page swaps 0
      Number of active context switches 0
      Number of passive context switches 0
      Block input operand 0
      Block output operand 96
      
82; * \ '; * "; * /;
83
84
85; * \ '; * "; * /;
86% put% upcase (e3969440a681a2408885998500000008);

**Comment by tomweber-sas** [comment]
Hey, glad you got the new one installed now. I think it's probably because the version number isn't different, so pip just didn't replace it that first time. uninstall and install is a good clean way that works well.

Hmmm, so I've done a lot of NLS work in saspy to support transcoding between utf-8 Python and all of the different encodings SAS supports. What you have in your config seem good; you're using the sas_zh script, which should run the correct sasv9cfg file to set the encoding and all of that correctly, so that looks good. The encoding values you used for the 'encoding' key are right for that: sas euc-cn matched python gb2312 or euc-cn, or a number of other aliases.

Ok, I tried to running this to see what I saw and it worked fine. But, I was running against v940m5, which has a bug in it (hoping for a hotfix soon), so I have 2 paths in my code for sd2df, because the 'better' code gen is broken with that bug, so I use a slightly different set of code that's not as good, but works. When I tried this again against v940m4, and went through the 'better' path, I see the error on the filename statement and the step fails. I see that SAS doesn't allow 'binary' encoing on the filename, apparently because this is a multibyte session? I'll have to check into that, but that seems to be the case. So that just plain fails.

Here's my run, which I have to guess is the same as what you got (it's helpful when you show all 
the code and output you got for a problem :) ). 
```
>>> import saspy
>>> sas = saspy.SASsession(cfgname='tdi_cn')
SAS Connection established. Subprocess id is 28783

No encoding value provided. Will try to determine the correct encoding.
Setting encoding to gb2312 based upon the SAS session encoding value of euc-cn.

>>> sas
Access Method         = STDIO
SAS Config name       = tdi_cn
SAS Config file       = /opt/tom/github/saspy/saspy/sascfg_personal.py
WORK Path             = /usr/tmp/SAS_workF0100000706F_tom64-5/
SAS Version           = 9.04.01M4P11092016
SASPy Version         = 3.3.5
Teach me SAS          = False
Batch                 = False
Results               = Pandas
SAS Session Encoding  = euc-cn
Python Encoding value = gb2312
SAS process Pid value = 28783


>>> cars = sas.sasdata('cars','sashelp', results='text')
>>> cars.head()
                                                              SAS 系统                    2020年 4月22日 星期三 上午09时29分37秒   1

                                       Drive                   Engine                                 MPG_
 Obs Make  Model          Type  Origin Train     MSRP  Invoice  Size  Cylinders Horsepower MPG_City Highway Weight Wheelbase Length

   1 Acura MDX            SUV    Asia  All    $36,945  $33,337   3.5      6         265       17       23    4451     106      189
   2 Acura RSX Type S 2dr Sedan  Asia  Front  $23,820  $21,761   2.0      4         200       24       31    2778     101      172
   3 Acura TSX 4dr        Sedan  Asia  Front  $26,990  $24,647   2.4      4         200       22       29    3230     105      183
   4 Acura TL 4dr         Sedan  Asia  Front  $33,195  $30,299   3.2      6         270       20       28    3575     108      186
   5 Acura 3.5 RL 4dr     Sedan  Asia  Front  $43,755  $39,014   3.5      6         225       18       24    3880     115      197
>>> df = cars.to_df()
error occured in SAS during sasdata2dataframe. Trying to return the saslog instead of a data frame.
>>> print(df)

97   filename sock socket ':52169' lrecl=1 recfm=f encoding=binary;
97   filename sock socket ':52169' lrecl=1 recfm=f encoding=binary;
                                                            ------
                                                            24
ERROR 24-2: 对于“encoding”选项无效的值。

ERROR: FILENAME 语句出错。
98   data _null_; set sashelp.'cars'n ;
99   format 'MSRP'n best32.; format 'Invoice'n best32.; format 'EngineSize'n best32.; format 'Cylinders'n best32.; format
99 ! 'Horsepower'n best32.; format 'MPG_City'n best32.;
100  format 'MPG_Highway'n best32.; format 'Weight'n best32.; format 'Wheelbase'n best32.; format 'Length'n best32.; file sock; put
100! 'Make'n '02'x;
101  put 'Model'n '02'x; put 'Type'n '02'x; put 'Origin'n '02'x; put 'DriveTrain'n '02'x; put 'MSRP'n '02'x; put 'Invoice'n '02'x;
101! put 'EngineSize'n '02'x; put 'Cylinders'n '02'x; put 'Horsepower'n '02'x; put 'MPG_City'n '02'x;
102  put 'MPG_Highway'n '02'x; put 'Weight'n '02'x; put 'Wheelbase'n '02'x; put 'Length'n '01'x; run;
NOTE: FILE SOCK 是:
      文件名=/opt/tom/github/saspy/saspy/sock.dat,
      所有者名=sastpw,组名称=r&d,访问权限=-rw-r--r--,
      上次修改时间=2020年04月22日 09时29分53秒

NOTE: 有 6420 条记录写入到 FILE SOCK 中。
      最小记录长度是 3。
      最大记录长度是 41。
NOTE: 从数据集 SASHELP.CARS. 读取了 428 个观测
NOTE: “DATA 语句”所用时间（总处理时间）:
      实际时间          0.04 秒
      CPU 时间          0.01 秒

103
```
I believe I have a workaround for you, and I will investigate this and try to see about fixing it for real in saspy. Given this works with the alternate code (for m5), you can set an attribute on the SASsession to force that code path:
```
sas.m5dsbug=True
```
You can set that in your config file too; so it's set all the time:
```
ssh48 = {'saspath' : ...
         'm5dsbug' : True,
         ...}
```
Here's the run after setting that, which works like the first time I tried it against m5:
```
104
105
106
>>> sas.m5dsbug
False
>>> sas.m5dsbug=True
>>> df = cars.to_df()
>>> df.head()
    Make           Model   Type Origin DriveTrain   MSRP  Invoice  EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  Weight  Wheelbase  Length
0  Acura             MDX    SUV   Asia        All  36945    33337         3.5        6.0         265        17           23    4451        106     189
1  Acura  RSX Type S 2dr  Sedan   Asia      Front  23820    21761         2.0        4.0         200        24           31    2778        101     172
2  Acura         TSX 4dr  Sedan   Asia      Front  26990    24647         2.4        4.0         200        22           29    3230        105     183
3  Acura          TL 4dr  Sedan   Asia      Front  33195    30299         3.2        6.0         270        20           28    3575        108     186
4  Acura      3.5 RL 4dr  Sedan   Asia      Front  43755    39014         3.5        6.0         225        18           24    3880        115     197
>>> 
```
Can you set sas.m5dsbug=True in your config definition or after you create your SASsession and see if this works right for your cases? 

Thanks!
Tom

**Comment by jinwzhang** [comment]
Thanks Tom for your kind help,

I will try this method soon.

Beside that, I had figgured out  way to walk around this issue. Which is use sas_en client.

ssh48      = {'saspath' : '/opt/sas94/SASHome/SASFoundation/9.4/bin/sas_en',
             ...
            'encoding': 'gb2312',
            }
then use: sas._io.sascfg.encoding='override_encoding'.

I had put this new feature with team production, and it is a very useful update.

Thank you Tom, and I think issue is closed well.

Have a good one.

Jim

**Comment by tomweber-sas** [comment]
Hey Jim.

Just let me know if you need anything else. I'll look into fixing the m5dsbug issue for the multibyte languages, so you don't need to specify that in the future. FWIW, you don't need to specify the 'ecoding' at all. I will  set it correctly based upon the SAS session encoding.
From the saspy Doc:
```
encoding -
NOTE: as of saspy V2.4.2, you no longer need to set the encoding. SASpy will determine the
SAS session encoding and map that to the Python encoding for you.
```
In fact the only reason to se it it to get rid of the message I print telling you what I set it to:
```
>>> sas = saspy.SASsession()
SAS Connection established. Subprocess id is 8830

No encoding value provided. Will try to determine the correct encoding.
Setting encoding to gb2312 based upon the SAS session encoding value of euc-cn.
```
Even for the case where you specifyit, but it's the wrong one, I overide what you had to set it correctly, so you really don't ever need to change it to other than what saspy set's it correctly to.
```
>>> sas = saspy.SASsession(cfgname='tdi_cn', encoding='latin1')
SAS Connection established. Subprocess id is 8927

The encoding value provided doesn't match the SAS session encoding.
SAS encoding is euc-cn. Specified encoding is latin1.
Using encoding gb2312 instead to avoid transcoding problems.
You can override this change, if you think you must, by changing the encoding attribute of the SASsession object, as follows.
If you had 'sas = saspy.SASsession(), then submit: "sas._io.sascfg.encoding='override_encoding'" to change it.

>>> sas
Access Method         = STDIO
SAS Config name       = tdi_cn
SAS Config file       = /opt/tom/github/saspy/saspy/sascfg_personal.py
WORK Path             = /usr/tmp/SAS_workFEFC000022DF_tom64-5/
SAS Version           = 9.04.01M4P11092016
SASPy Version         = 3.3.5
Teach me SAS          = False
Batch                 = False
Results               = Pandas
SAS Session Encoding  = euc-cn
Python Encoding value = gb2312
SAS process Pid value = 8927
```
You should be ok running the sas_zh version, just nee to set
'm5dsbug' : True.
In your config for that one till I address this with a fix; which will have to be making it run that same codepath for the mutlibyte cases since SAS won't allow a binary encoding on the socket filename in that case, which is too bad, but it should all still work right.

Thanks,
Tom

**Comment by tomweber-sas** [comment]
And, just one last thing I noticed so far, is that 'binary' seems to be allowed running sas_zh, it you run with encoding=utf8 instead of euc-cn. Don't know if that's an option for you, but if so, then you don't need m5dsbug=True.

```
>>> import saspy; sas = saspy.SASsession(cfgname='tdi_cn'); sas
SAS Connection established. Subprocess id is 10094

No encoding value provided. Will try to determine the correct encoding.
Setting encoding to utf_8 based upon the SAS session encoding value of utf-8.

Access Method         = STDIO
SAS Config name       = tdi_cn
SAS Config file       = /opt/tom/github/saspy/saspy/sascfg_personal.py
WORK Path             = /usr/tmp/SAS_workD9B90000276E_tom64-5/
SAS Version           = 9.04.01M4P11092016
SASPy Version         = 3.3.5
Teach me SAS          = False
Batch                 = False
Results               = Pandas
SAS Session Encoding  = utf-8
Python Encoding value = utf_8
SAS process Pid value = 10094

>>> cars = sas.sd2df('cars','sashelp')
>>> print(sas.lastlog())

68   filename sock socket ':59220' lrecl=1 recfm=f encoding=binary;
69   data _null_; set sashelp.'cars'n ;
70   format 'MSRP'n best32.; format 'Invoice'n best32.; format 'EngineSize'n best32.; format 'Cylinders'n best32.; format
70 ! 'Horsepower'n best32.; format 'MPG_City'n best32.;
71   format 'MPG_Highway'n best32.; format 'Weight'n best32.; format 'Wheelbase'n best32.; format 'Length'n best32.; file sock; put
71 ! 'Make'n '02'x;
72   put 'Model'n '02'x; put 'Type'n '02'x; put 'Origin'n '02'x; put 'DriveTrain'n '02'x; put 'MSRP'n '02'x; put 'Invoice'n '02'x;
72 ! put 'EngineSize'n '02'x; put 'Cylinders'n '02'x; put 'Horsepower'n '02'x; put 'MPG_City'n '02'x;
73   put 'MPG_Highway'n '02'x; put 'Weight'n '02'x; put 'Wheelbase'n '02'x; put 'Length'n '01'x; run;
NOTE: FILE SOCK 是:
      Local Host Name=tom64-5,
      Local Host IP addr=127.0.0.1,
      Peer Hostname Name=localhost,
      Peer IP addr=127.0.0.1,Peer Name=N/A,
      Peer Portno=59220,Lrecl=1,Recfm=Fixed

NOTE: 有 34167 条记录写入到 FILE SOCK 中。
NOTE: 从数据集 SASHELP.CARS. 读取了 428 个观测
NOTE: “DATA 语句”所用时间（总处理时间）:
      实际时间          0.13 秒
      CPU 时间          0.07 秒

74
75
76
77
>>> cars
           Make                             Model    Type  Origin DriveTrain   MSRP  Invoice  EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  Weight  Wheelbase  Length
0         Acura                               MDX     SUV    Asia        All  36945    33337         3.5        6.0         265        17           23    4451        106     189
1         Acura                    RSX Type S 2dr   Sedan    Asia      Front  23820    21761         2.0        4.0         200        24           31    2778        101     172
2         Acura                           TSX 4dr   Sedan    Asia      Front  26990    24647         2.4        4.0         200        22           29    3230        105     183
3         Acura                            TL 4dr   Sedan    Asia      Front  33195    30299         3.2        6.0         270        20           28    3575        108     186
4         Acura                        3.5 RL 4dr   Sedan    Asia      Front  43755    39014         3.5        6.0         225        18           24    3880        115     197
[...]
```

tdi_cn   = {'saspath' : '/tdi/mva-v940m4/usrlibsas/laxno/SASFoundation/9.4/bin/sas_zh',
            'options' : ["-encoding", "utf8"]
            }
```

Tom

**Comment by tomweber-sas** [comment]
@jinwzhang I've dug into this problem where encoding=binary is sometimes not allowed on the socket access method, which is causing this failure. It turns out that it is only in the DBCS builds of SAS and only when the encoding isn't UTF-8. There's really no reason it can't send bytes in that case, but that's what's happening and isn't going to change. So, I've been trying some different settings to try to get the same behavior as this for both the existing cases that work and for this case with DBCS and local encodings (not utf8). I've got a fix that appears to be working as expected for all cases now.

I'm hoping you can verify that it does fix your case, as well as what I've seen here trying to reproduce it. This same error was also happening with upload/download in the environment I was able to use to reproduce the error, so it was even more pervasive than just sd2df - and the upload/download didn't happen to have the m5dsbug alternate code path, so those appear broke with no workaround.

I've pushed this fix to master and was hoping you could download master and uninstall/install and see if this does fix the problem for you. You should see the same error with upload and download before getting master too, and that should then also be fixed.

Is validating this something you can do?

Thanks!
Tom

**Comment by tomweber-sas** [comment]
These fixes are now in the new current release, V3.3.6. If you have any other issues or questions, just let me know!

Thanks,
Tom
